# Déjame escuchar la música

<h1>Contenido<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Introducción-" data-toc-modified-id="Introducción--1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Introducción <a id="intro"></a></a></span><ul class="toc-item"><li><span><a href="#Objetivo:" data-toc-modified-id="Objetivo:-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Objetivo:</a></span></li><li><span><a href="#Etapas" data-toc-modified-id="Etapas-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Etapas</a></span></li></ul></li><li><span><a href="#Etapa-1.-Descripción-de-los-datos-" data-toc-modified-id="Etapa-1.-Descripción-de-los-datos--2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Etapa 1. Descripción de los datos <a id="data_review"></a></a></span><ul class="toc-item"><li><span><a href="#Escribe-observaciones-de-tu-parte.-Estas-son-algunas-de-las-preguntas-que-pueden-ser-útiles:-" data-toc-modified-id="Escribe-observaciones-de-tu-parte.-Estas-son-algunas-de-las-preguntas-que-pueden-ser-útiles:--2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles: <a id="data_review_conclusions"></a></a></span></li></ul></li><li><span><a href="#Etapa-2.-Preprocesamiento-de-datos-" data-toc-modified-id="Etapa-2.-Preprocesamiento-de-datos--3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Etapa 2. Preprocesamiento de datos <a id="data_preprocessing"></a></a></span><ul class="toc-item"><li><span><a href="#Estilo-del-encabezado-" data-toc-modified-id="Estilo-del-encabezado--3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Estilo del encabezado <a id="header_style"></a></a></span></li><li><span><a href="#Valores-ausentes-" data-toc-modified-id="Valores-ausentes--3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Valores ausentes <a id="missing_values"></a></a></span></li><li><span><a href="#Duplicados-" data-toc-modified-id="Duplicados--3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Duplicados <a id="duplicates"></a></a></span></li><li><span><a href="#Tus-observaciones-" data-toc-modified-id="Tus-observaciones--3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Tus observaciones <a id="data_preprocessing_conclusions"></a></a></span></li></ul></li><li><span><a href="#Etapa-3.-Prueba-de-hipótesis-" data-toc-modified-id="Etapa-3.-Prueba-de-hipótesis--4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Etapa 3. Prueba de hipótesis <a id="hypothesis"></a></a></span><ul class="toc-item"><li><span><a href="#Hipótesis:-comparar-el-comportamiento-del-usuario-o-la-usuaria-en-las-dos-ciudades-" data-toc-modified-id="Hipótesis:-comparar-el-comportamiento-del-usuario-o-la-usuaria-en-las-dos-ciudades--4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id="activity"></a></a></span></li><li><span><a href="#Nota" data-toc-modified-id="Nota-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Nota</a></span></li></ul></li></ul></div>

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [1]:
# Importar pandas
import pandas as pd

Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
# Leer el archivo y almacenarlo en df
df = pd.read_csv("/datasets/music_project_en.csv")

Muestra las 10 primeras filas de la tabla:

In [3]:
# Obtener las 10 primeras filas de la tabla df
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [4]:
# Obtener la información general sobre nuestros datos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. El tercer problema el nombre de las columnas, ya que `'userID'` que deberia ser `'user_id'` para cuando renombremos todos los encabezados a minusculas pa que no se confunda como una sola palabra.

Las columnas columnas `'City'`,`'time'` y `'Day'` no tienen un nombre que describa muy bien los valores que contienen esas columnas, la columna `'City'` deberia ser `'city_user'` ya que es la ciudad del usuario que escucha la cancion, la columna `'time'` deberia ser `'time_play'` ya que es la hora exacta en que se comenzo a reproducir la cancion y la columna `'Day'` deberia ser `'day_play'` ya que es el dia que se reprodujo la cancion en la semana.






`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`

     Tenemos datos que hacen referencia a informacion de las canciones, esta informacion tiene columnas importantes como 'City' y 'Day' que nos permiten saber si la actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad, ademas de esta informacion podemos generar otras hipotesis para saber si en realidad la actividad de los usuarios y usuarias difiere en realidad por los generos de las canciones y  mas importante aun si su actividad se centra en una hora concreta del dia y entendemos que podemos responder estas hipotesis por que las columnas tienen nombre referentes a los datos que almacenan, aun que algunos son pocos descriptivos en los valores que almacenan.

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`
       
       Si hay suficiente informacion ya que no solo podemos responder la hipotesis objetiva si no tambien podemos saber si en realidad no difiere la actividad de los usuarios solo por la ciudad y el dia, si no tambien por la hora del dia y genero de las canciones

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`
       
       En las primeras 10 filas obtenidas de la tabla podemos ver que en la fila del indice 9 tiene un dato NaN en la columna artista

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corregimos el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [5]:
# Muestra los nombres de las columnas
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [6]:
# Bucle en los encabezados poniendo todo en minúsculas
col_names_lower=[]

for old_name_column in df.columns:
    name_column_lowered=old_name_column.lower()
    col_names_lower.append(name_column_lowered)

df.columns=col_names_lower
print(df.columns)

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')


Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [7]:
# Bucle en los encabezados eliminando los espacios
col_names_strip=[]

for old_name_column in df.columns:
    name_column_stripped=old_name_column.strip()
    col_names_strip.append(name_column_stripped)

df.columns=col_names_strip
print(df.columns)

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [8]:
# Cambiar el nombre de la columna "userid"
df.rename(columns={"userid":"user_id"},inplace=True)
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Comprueba el resultado. Muestra los encabezados una vez más:

In [9]:
# Comprobar el resultado: la lista de encabezados
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [10]:
# Calcular el número de valores ausentes
print(df.isna().sum())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [11]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
columns_to_replace=["track","artist","genre"]

for column in columns_to_replace:
    df[column].fillna("unknown",inplace=True)


Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [12]:
# Contar valores ausentes
print(df.isna().sum())

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [13]:
# Contar duplicados explícitos
print(df.duplicated().sum())

3826


Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [14]:
# Eliminar duplicados explícitos
df.drop_duplicates(inplace=True)

Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [15]:
# Comprobar de nuevo si hay duplicados

print(df.duplicated().sum())

0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [16]:
# Inspeccionar los nombres de géneros únicos
column_genre=df["genre"].sort_values()
print(column_genre.unique())
print(column_genre.nunique())

['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [17]:
# Función para reemplazar duplicados implícitos
def replace_wrong_genres(df,wrong_genres,correct_genre):
    for wrong_genre in wrong_genres:
        df["genre"]=df["genre"].replace(wrong_genre,correct_genre)
    return df

Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [18]:
# Eliminar duplicados implícitos
genres_duplicates=["hip","hop","hip-hop"]
correct_genre="hiphop"
df=replace_wrong_genres(df,genres_duplicates,correct_genre)

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [19]:
# Comprobación de duplicados implícitos
column_genre=df["genre"].sort_values()
print(column_genre.unique())
print(column_genre.nunique())

['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

He notado que cuando se analiza una gran cantidad de duplicados como fue el caso de los generos, es dificil encontrar los generlos duplicados implicitos y mal escritos a la vista, osea leerlos uno por uno, aun que si ordenas la lista con el metodo sort_values() simplifica un poco la tarea, ya identificados es facil analizar y comparar si fueron corregidos utilizando el metodo nunique() ya que puedes contar los datos y si fueron reducidos es por que se realizo una correccion exitosa.

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [20]:
# Contar las canciones reproducidas en cada ciudad
print(df.groupby(by="city")["city"].count())

city
Shelbyville    18512
Springfield    42741
Name: city, dtype: int64



En la ciudad de Shelbyville se han reproducido 18512 canciones.
En la ciudad de Srpingfield se han reproducido 42741 canciones.

En la sintaxis del `df.groupby(by='....')['column'].method()` la columna utilizada en 'column' no importa ya que el numero de filas es el mismo en todas por lo que el conteo total no cambia.

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [21]:
# Calcular las canciones reproducidas en cada uno de los tres días
print(df.groupby(by="day")["day"].count())

day
Friday       21840
Monday       21354
Wednesday    18059
Name: day, dtype: int64



El dia Lunes de Shelbyville se han reproducido 21840 canciones.
El día Miercoles se han reproducido 18059 canciones.
El día Viernes se han reproducido 21840 canciones


Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [22]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
def number_tracks(df,day,city):
    
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    filter_day=df[df["day"]==day]
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    filter_city=filter_day[filter_day["city"]==city]
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    count_user_id=filter_city["user_id"].count()
    # Devolve el número de valores de la columna 'user_id'
    return count_user_id

Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [23]:
# El número de canciones reproducidas en Springfield el lunes
print(number_tracks(df,"Monday","Springfield"))

15740


In [25]:
# El número de canciones reproducidas en Shelbyville el lunes
print(number_tracks(df,"Monday","Shelbyville"))

5614


In [24]:
# El número de canciones reproducidas en Springfield el miércoles
print(number_tracks(df,"Wednesday","Springfield"))

11056


In [25]:
# El número de canciones reproducidas en Shelbyville el miércoles
print(number_tracks(df,"Wednesday","Shelbyville"))

7003


In [26]:
# El número de canciones reproducidas en Springfield el viernes
print(number_tracks(df,"Friday","Springfield"))

15945


In [27]:
# El número de canciones reproducidas en Shelbyville el viernes
print(number_tracks(df,"Friday","Shelbyville"))

5895


**Conclusiones**


La hipotesis es parcialmente correcta ya que en los 3 dias analizados Springfield supera en canciones reproducidas a Shelbyville pero en ningun día ambas ciudades coinciden en conseguir el numero maximo de reproducciones, por lo que podemos decir la actividad de los usuarios y las usuarias solo difiere dependiendo de la ciudad y no del día.



[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>



El dataset del ejercicio permite responde al objetivo de probar la hipotesis:

    -La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.

Pero el dataset permite responder otras hipotesis ya que la actividad de los usuarios y usuarias puede diferir de otros factores como el genero de la cancion y la hora que se comenzo a reproducir la cancion, por ejemplo:

    -Si los usuarios tienen mas actividad en la mañana, tarde o noche de cada dia.
    -Si los usuarios tienen mas actividad escuchando cierto genero dependiendo del día, como sucede en días de fiesta como es el viernes o dias de trabajo como lunes y miercoles.
    -Si los usuarios de una ciudad tienen mas actividad con un genero y/o un horario especifico.

Permitiendo obtener informacion valiosa para la empresa que entrega el servicio de musica permitiendo distintas acciones como:
    
    -Realizar eventos en ciudades que escuchan mucho un genero para recaudar dinero.
    -Buscar más artistas en e para aprovechar la actividad de los usuarios que escuchan ese genero.

[Volver a Contenidos](#back)